File: Sources/AstrolabeProtocol/Framing/RuntimeFrameCodec.swift
//
//  RuntimeFrameCodec.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/10.
//

import Foundation

public struct RuntimeFrameCodec: Sendable {
    /// Default upper bound for one framed JSON payload.
    public static let defaultMaximumPayloadSize = 16 * 1024 * 1024

    /// Maximum number of payload bytes accepted in a single frame.
    public let maximumPayloadSize: Int

    public init() {
        maximumPayloadSize = Self.defaultMaximumPayloadSize
    }

    public init(maximumPayloadSize: Int) throws {
        guard maximumPayloadSize > 0,
              maximumPayloadSize <= Int(UInt32.max) else {
            throw RuntimeFrameError.invalidMaximumPayloadSize(maximumPayloadSize)
        }
        self.maximumPayloadSize = maximumPayloadSize
    }

    public func encode(payload: Data) throws -> Data {
        guard !payload.isEmpty else {
            throw RuntimeFrameError.emptyPayload
        }
        guard payload.count <= maximumPayloadSize else {
            throw RuntimeFrameError.payloadTooLarge(
                actual: payload.count,
                maximum: maximumPayloadSize
            )
        }

        let length = UInt32(payload.count)
        var frame = Data(capacity: MemoryLayout<UInt32>.size + payload.count)
        frame.append(UInt8((length >> 24) & 0xFF))
        frame.append(UInt8((length >> 16) & 0xFF))
        frame.append(UInt8((length >> 8) & 0xFF))
        frame.append(UInt8(length & 0xFF))
        frame.append(payload)
        return frame
    }

    public func makeStreamDecoder() -> RuntimeFrameStreamDecoder {
        RuntimeFrameStreamDecoder(maximumPayloadSize: maximumPayloadSize)
    }
}

public struct RuntimeFrameStreamDecoder: Sendable {
    /// Number of bytes retained while waiting for a complete frame.
    public var pendingByteCount: Int {
        buffer.count
    }

    private let maximumPayloadSize: Int
    private var buffer = Data()

    fileprivate init(maximumPayloadSize: Int) {
        self.maximumPayloadSize = maximumPayloadSize
    }

    public mutating func append(_ data: Data) throws -> [Data] {
        buffer.append(data)
        var payloads = [Data]()

        while buffer.count >= MemoryLayout<UInt32>.size {
            let payloadLength = Int(readPayloadLength())
            guard payloadLength > 0 else {
                throw RuntimeFrameError.emptyPayload
            }
            guard payloadLength <= maximumPayloadSize else {
                throw RuntimeFrameError.payloadTooLarge(
                    actual: payloadLength,
                    maximum: maximumPayloadSize
                )
            }

            let frameLength = MemoryLayout<UInt32>.size + payloadLength
            guard buffer.count >= frameLength else {
                break
            }

            payloads.append(
                buffer.subdata(in: MemoryLayout<UInt32>.size ..< frameLength)
            )
            buffer.removeSubrange(0 ..< frameLength)
        }

        return payloads
    }

    public mutating func reset() {
        buffer.removeAll(keepingCapacity: true)
    }

    private func readPayloadLength() -> UInt32 {
        buffer.prefix(MemoryLayout<UInt32>.size).reduce(0) { partialResult, byte in
            (partialResult << 8) | UInt32(byte)
        }
    }
}

public enum RuntimeFrameError: Error, Equatable, Sendable {
    case emptyPayload
    case invalidMaximumPayloadSize(Int)
    case payloadTooLarge(actual: Int, maximum: Int)
}

File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/framing/RuntimeFrameCodec.kt
//
//  RuntimeFrameCodec.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

/** Encodes and decodes length-prefixed Astrolabe wire frames. */
public class RuntimeFrameCodec(
    /** Maximum number of payload bytes accepted in one frame. */
    public val maximumPayloadSize: Int = DEFAULT_MAXIMUM_PAYLOAD_SIZE
) {
    init {
        require(maximumPayloadSize > 0) {
            "maximumPayloadSize must be greater than zero"
        }
    }

    /** Creates one network-byte-order frame for [payload]. */
    public fun encode(payload: ByteArray): ByteArray {
        if (payload.isEmpty()) {
            throw RuntimeFrameException.EmptyPayload
        }
        if (payload.size > maximumPayloadSize) {
            throw RuntimeFrameException.PayloadTooLarge(
                actual = payload.size.toLong(),
                maximum = maximumPayloadSize
            )
        }

        val length = payload.size
        return ByteArray(FRAME_HEADER_SIZE + length).also { frame ->
            frame[0] = (length ushr 24).toByte()
            frame[1] = (length ushr 16).toByte()
            frame[2] = (length ushr 8).toByte()
            frame[3] = length.toByte()
            payload.copyInto(frame, destinationOffset = FRAME_HEADER_SIZE)
        }
    }

    /** Creates a stateful decoder for fragmented byte streams. */
    public fun makeStreamDecoder(): RuntimeFrameStreamDecoder =
        RuntimeFrameStreamDecoder(maximumPayloadSize)

    public companion object {
        /** Default upper bound for one framed JSON payload. */
        public const val DEFAULT_MAXIMUM_PAYLOAD_SIZE: Int = 16 * 1024 * 1024

        internal const val FRAME_HEADER_SIZE: Int = 4
    }
}

/** Decodes zero or more complete frames from fragmented input. */
public class RuntimeFrameStreamDecoder internal constructor(
    private val maximumPayloadSize: Int
) {
    private val header = ByteArray(RuntimeFrameCodec.FRAME_HEADER_SIZE)
    private var headerByteCount: Int = 0
    private var payload: ByteArray? = null
    private var payloadByteCount: Int = 0

    /** Number of bytes retained while waiting for a complete frame. */
    public val pendingByteCount: Int
        get() = headerByteCount + payloadByteCount

    /** Appends bytes and returns every complete payload now available. */
    public fun append(data: ByteArray): List<ByteArray> {
        val payloads = mutableListOf<ByteArray>()
        var offset = 0
        while (offset < data.size) {
            if (payload == null) {
                val headerBytesToCopy = minOf(
                    RuntimeFrameCodec.FRAME_HEADER_SIZE - headerByteCount,
                    data.size - offset
                )
                data.copyInto(
                    destination = header,
                    destinationOffset = headerByteCount,
                    startIndex = offset,
                    endIndex = offset + headerBytesToCopy
                )
                headerByteCount += headerBytesToCopy
                offset += headerBytesToCopy
                if (headerByteCount < RuntimeFrameCodec.FRAME_HEADER_SIZE) {
                    continue
                }
                payload = ByteArray(validatedPayloadLength())
            }

            val currentPayload = payload ?: continue
            val payloadBytesToCopy = minOf(
                currentPayload.size - payloadByteCount,
                data.size - offset
            )
            data.copyInto(
                destination = currentPayload,
                destinationOffset = payloadByteCount,
                startIndex = offset,
                endIndex = offset + payloadBytesToCopy
            )
            payloadByteCount += payloadBytesToCopy
            offset += payloadBytesToCopy
            if (payloadByteCount == currentPayload.size) {
                payloads += currentPayload
                resetFrameState()
            }
        }
        return payloads
    }

    /** Discards any incomplete frame bytes. */
    public fun reset() {
        resetFrameState()
    }

    private fun validatedPayloadLength(): Int {
        val payloadLength = readPayloadLength()
        if (payloadLength == 0L) {
            throw RuntimeFrameException.EmptyPayload
        }
        if (payloadLength > maximumPayloadSize.toLong()) {
            throw RuntimeFrameException.PayloadTooLarge(
                actual = payloadLength,
                maximum = maximumPayloadSize
            )
        }
        return payloadLength.toInt()
    }

    private fun resetFrameState() {
        headerByteCount = 0
        payload = null
        payloadByteCount = 0
    }

    private fun readPayloadLength(): Long =
        (0 until RuntimeFrameCodec.FRAME_HEADER_SIZE).fold(0L) { length, index ->
            (length shl 8) or (header[index].toLong() and 0xFF)
        }
}

/** Failures produced while reading or writing a wire frame. */
public sealed class RuntimeFrameException(message: String) : Exception(message) {
    /** The protocol does not permit an empty frame payload. */
    public data object EmptyPayload : RuntimeFrameException("Frame payload cannot be empty")

    /** A declared or encoded payload exceeds the configured limit. */
    public class PayloadTooLarge(
        /** Number of payload bytes declared or supplied. */
        public val actual: Long,
        /** Maximum number of payload bytes accepted. */
        public val maximum: Int
    ) : RuntimeFrameException("Frame payload size $actual exceeds maximum $maximum")
}
